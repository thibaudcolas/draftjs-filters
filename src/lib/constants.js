// @flow
export const ATOMIC = "atomic"
export const UNSTYLED = "unstyled"
export const UNORDERED_LIST_ITEM = "unordered-list-item"
export const ORDERED_LIST_ITEM = "ordered-list-item"

export const IMAGE = "IMAGE"
