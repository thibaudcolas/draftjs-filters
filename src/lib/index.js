// @flow
import {
  preserveAtomicBlocks,
  resetAtomicBlocks,
  removeInvalidAtomicBlocks,
} from "./filters/atomic"
import {
  removeInvalidDepthBlocks,
  limitBlockDepth,
  filterBlockTypes,
} from "./filters/blocks"
import { filterInlineStyles } from "./filters/styles"
import {
  filterEntityRanges,
  shouldKeepEntityType,
  shouldRemoveImageEntity,
  shouldKeepEntityByAttribute,
  filterEntityData,
} from "./filters/entities"
import { replaceTextBySpaces } from "./filters/text"
import { filterEditorState } from "./filters/editor"

export {
  preserveAtomicBlocks,
  resetAtomicBlocks,
  removeInvalidAtomicBlocks,
  removeInvalidDepthBlocks,
  limitBlockDepth,
  filterBlockTypes,
  filterInlineStyles,
  filterEntityRanges,
  shouldKeepEntityType,
  shouldRemoveImageEntity,
  shouldKeepEntityByAttribute,
  filterEntityData,
  replaceTextBySpaces,
  filterEditorState,
}
