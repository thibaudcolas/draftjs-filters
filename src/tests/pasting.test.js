import { EditorState, convertFromRaw, convertToRaw } from "draft-js"
import snapshotDiff from "snapshot-diff"

import { filterEditorState } from "../lib/index"

// https://github.com/jest-community/snapshot-diff
expect.addSnapshotSerializer(snapshotDiff.getSnapshotDiffSerializer())

const data = {
  "applepages-chrome62-macos1013": require("../../pasting/applepages-chrome62-macos1013.json"),
  "applepages-firefox57-macos1013": require("../../pasting/applepages-firefox57-macos1013.json"),
  "applepages-safari-ios11": require("../../pasting/applepages-safari-ios11.json"),
  "applepages-safari11-macos1013": require("../../pasting/applepages-safari11-macos1013.json"),
  "dropboxpaper-chrome62-macos1013": require("../../pasting/dropboxpaper-chrome62-macos1013.json"),
  "dropboxpaper-chrome62-win81": require("../../pasting/dropboxpaper-chrome62-win81.json"),
  "dropboxpaper-edge16-win10": require("../../pasting/dropboxpaper-edge16-win10.json"),
  "dropboxpaper-firefox57-macos1013": require("../../pasting/dropboxpaper-firefox57-macos1013.json"),
  "dropboxpaper-firefox57-win81": require("../../pasting/dropboxpaper-firefox57-win81.json"),
  "dropboxpaper-ie11-unsupported-strippastedstyles-win81": require("../../pasting/dropboxpaper-ie11-unsupported-strippastedstyles-win81.json"),
  "dropboxpaper-ie11-unsupported-win81": require("../../pasting/dropboxpaper-ie11-unsupported-win81.json"),
  "dropboxpaper-safari-ios11": require("../../pasting/dropboxpaper-safari-ios11.json"),
  "dropboxpaper-safari11-macos1013": require("../../pasting/dropboxpaper-safari11-macos1013.json"),
  "googledocs-chrome62-macos1013": require("../../pasting/googledocs-chrome62-macos1013.json"),
  "googledocs-chrome62-win81": require("../../pasting/googledocs-chrome62-win81.json"),
  "googledocs-edge16-win10": require("../../pasting/googledocs-edge16-win10.json"),
  "googledocs-firefox57-macos1013": require("../../pasting/googledocs-firefox57-macos1013.json"),
  "googledocs-firefox57-win81": require("../../pasting/googledocs-firefox57-win81.json"),
  "googledocs-ie11-strippastedstyles-win81": require("../../pasting/googledocs-ie11-strippastedstyles-win81.json"),
  "googledocs-ie11-win81": require("../../pasting/googledocs-ie11-win81.json"),
  "googledocs-safari-ios11": require("../../pasting/googledocs-safari-ios11.json"),
  "googledocs-safari11-macos1013": require("../../pasting/googledocs-safari11-macos1013.json"),
  "word-safari-ios11": require("../../pasting/word-safari-ios11.json"),
  "word2010-chrome62-win81": require("../../pasting/word2010-chrome62-win81.json"),
  "word2010-firefox57-win81": require("../../pasting/word2010-firefox57-win81.json"),
  "word2010-ie11-noequation-strippastedstyles-win81": require("../../pasting/word2010-ie11-noequation-strippastedstyles-win81.json"),
  "word2010-ie11-noequation-win81": require("../../pasting/word2010-ie11-noequation-win81.json"),
  "wordonline-chrome62-macos1013": require("../../pasting/wordonline-chrome62-macos1013.json"),
  "wordonline-chrome62-win81": require("../../pasting/wordonline-chrome62-win81.json"),
  "wordonline-edge16-win10": require("../../pasting/wordonline-edge16-win10.json"),
  "wordonline-firefox57-macos1013": require("../../pasting/wordonline-firefox57-macos1013.json"),
  "wordonline-firefox57-win81": require("../../pasting/wordonline-firefox57-win81.json"),
  "wordonline-ie11-win81": require("../../pasting/wordonline-ie11-win81.json"),
  "wordonline-safari11-macos1013": require("../../pasting/wordonline-safari11-macos1013.json"),
}

const FILTERS = {
  maxListNesting: 4,
  enableHorizontalRule: true,
  enableLineBreak: true,
  blockTypes: [
    "unstyled",
    "header-two",
    "header-three",
    "header-four",
    "unordered-list-item",
    "ordered-list-item",
  ],
  inlineStyles: ["BOLD", "ITALIC"],
  entityTypes: [
    {
      type: "LINK",
      attributes: ["url"],
      whitelist: {
        href: "^(?!#)",
      },
    },
    {
      type: "IMAGE",
      attributes: ["src"],
      whitelist: {
        src: "^http",
      },
    },
  ],
}

describe("Copy-paste samples", () => {
  it("word2010-chrome62-win81", () => {
    const raw = data["word2010-chrome62-win81"]
    const content = convertFromRaw(raw)
    const editorState = filterEditorState(
      Object.assign(
        {
          editorState: EditorState.createWithContent(content),
        },
        FILTERS,
      ),
    )

    const filtered = convertToRaw(editorState.getCurrentContent())

    // We snapshot the difference in the content before-after filtering, instead of the resulting content.
    // This makes it easier to review what exactly is filtered out.
    expect(snapshotDiff(raw, filtered)).toMatchSnapshot()
  })
})
