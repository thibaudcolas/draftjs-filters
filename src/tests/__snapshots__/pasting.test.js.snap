// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Copy-paste samples word2010-chrome62-win81 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -32,11 +32,11 @@
        "depth": 0,
        "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "2dc31",
        "text": "Inline styles",
-       "type": "header-one",
+       "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
@@ -64,31 +64,20 @@
          Object {
            "length": 11,
            "offset": 24,
            "style": "ITALIC",
          },
-         Object {
-           "length": 13,
-           "offset": 37,
-           "style": "STRIKETHROUGH",
-         },
        ],
        "key": "e1kg4",
        "text": "Unstyled, bold, italic, bold italic, strikethrough, monospace, unstyled, bold monospace",
        "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
-       "inlineStyleRanges": Array [
-         Object {
-           "length": 9,
-           "offset": 10,
-           "style": "UNDERLINE",
-         },
-       ],
+       "inlineStyleRanges": Array [],
        "key": "f3jvk",
        "text": "Unstyled, underline, bigger, smaller, superscript, subscript",
        "type": "unstyled",
      },
      Object {
@@ -99,15 +88,10 @@
          Object {
            "length": 7,
            "offset": 32,
            "style": "BOLD",
          },
-         Object {
-           "length": 20,
-           "offset": 53,
-           "style": "STRIKETHROUGH",
-         },
        ],
        "key": "d1736",
        "text": "Unstyled, red, text background, outline, small caps, double strikethrough, unstyled",
        "type": "unstyled",
      },
@@ -131,11 +115,11 @@
        "depth": 0,
        "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "5s4am",
        "text": "Inline entities",
-       "type": "header-one",
+       "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [
@@ -154,23 +138,13 @@
            "length": 12,
            "offset": 28,
          },
          Object {
            "key": 3,
-           "length": 13,
-           "offset": 42,
-         },
-         Object {
-           "key": 4,
            "length": 10,
            "offset": 57,
          },
-         Object {
-           "key": 5,
-           "length": 6,
-           "offset": 76,
-         },
        ],
        "inlineStyleRanges": Array [],
        "key": "cnvhf",
        "text": "Unstyled, link, email link, tooltip link, internal link, frame link, comment[T.C1] ",
        "type": "unstyled",
@@ -178,19 +152,14 @@
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [
          Object {
-           "key": 6,
+           "key": 4,
            "length": 17,
            "offset": 10,
          },
-         Object {
-           "key": 7,
-           "length": 6,
-           "offset": 27,
-         },
        ],
        "inlineStyleRanges": Array [
          Object {
            "length": 17,
            "offset": 10,
@@ -206,11 +175,11 @@
        "depth": 0,
        "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "bh960",
        "text": "Text blocks",
-       "type": "header-one",
+       "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
@@ -260,11 +229,11 @@
        "depth": 0,
        "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "6nap5",
        "text": "h1 - Heading one",
-       "type": "header-one",
+       "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
@@ -302,20 +271,20 @@
            "style": "BOLD",
          },
        ],
        "key": "adalk",
        "text": "h5 - heading five bold",
-       "type": "header-five",
+       "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "235sj",
        "text": "h6 - heading six",
-       "type": "header-six",
+       "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
@@ -433,36 +402,24 @@
        "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
-       "entityRanges": Array [
-         Object {
-           "key": 8,
-           "length": 1,
-           "offset": 0,
-         },
-       ],
+       "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "6umi7",
-       "text": "ðŸ“· Star list",
-       "type": "unstyled",
+       "text": " ",
+       "type": "atomic",
      },
      Object {
        "data": Object {},
        "depth": 0,
-       "entityRanges": Array [
-         Object {
-           "key": 9,
-           "length": 1,
-           "offset": 0,
-         },
-       ],
+       "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "fonu3",
-       "text": "ðŸ“· Star list",
-       "type": "unstyled",
+       "text": " ",
+       "type": "atomic",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
@@ -481,21 +438,15 @@
        "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
-       "entityRanges": Array [
-         Object {
-           "key": 10,
-           "length": 1,
-           "offset": 0,
-         },
-       ],
+       "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "nl29",
-       "text": "ðŸ“· Star list",
-       "type": "unstyled",
+       "text": " ",
+       "type": "atomic",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
@@ -518,11 +469,11 @@
        "depth": 0,
        "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "1l39f",
        "text": "Rich blocks",
-       "type": "header-one",
+       "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
@@ -547,21 +498,15 @@
        "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
-       "entityRanges": Array [
-         Object {
-           "key": 11,
-           "length": 1,
-           "offset": 0,
-         },
-       ],
+       "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "ff9kg",
-       "text": "ðŸ“·",
-       "type": "unstyled",
+       "text": " ",
+       "type": "atomic",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
@@ -580,21 +525,15 @@
        "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
-       "entityRanges": Array [
-         Object {
-           "key": 12,
-           "length": 1,
-           "offset": 0,
-         },
-       ],
+       "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "8cu2j",
-       "text": "ðŸ“·",
-       "type": "unstyled",
+       "text": " ",
+       "type": "atomic",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
@@ -728,11 +667,11 @@
        "depth": 0,
        "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "frsp9",
        "text": "Misc",
-       "type": "header-one",
+       "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
@@ -840,11 +779,11 @@
        "depth": 0,
        "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "dvkpk",
        "text": "Embeds",
-       "type": "header-one",
+       "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
@@ -860,42 +799,24 @@
        "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
-       "entityRanges": Array [
-         Object {
-           "key": 13,
-           "length": 1,
-           "offset": 0,
-         },
-       ],
+       "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "7ri7s",
-       "text": "ðŸ“· ",
-       "type": "unstyled",
+       "text": " ",
+       "type": "atomic",
      },
      Object {
        "data": Object {},
        "depth": 0,
-       "entityRanges": Array [
-         Object {
-           "key": 14,
-           "length": 1,
-           "offset": 0,
-         },
-       ],
-       "inlineStyleRanges": Array [
-         Object {
-           "length": 5,
-           "offset": 1,
-           "style": "BOLD",
-         },
-       ],
+       "entityRanges": Array [],
+       "inlineStyleRanges": Array [],
        "key": "alt7k",
-       "text": "ðŸ“·Shape",
-       "type": "unstyled",
+       "text": " ",
+       "type": "atomic",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
@@ -911,21 +832,15 @@
        "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
-       "entityRanges": Array [
-         Object {
-           "key": 15,
-           "length": 1,
-           "offset": 0,
-         },
-       ],
+       "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "4t12b",
-       "text": "ðŸ“·",
-       "type": "unstyled",
+       "text": " ",
+       "type": "atomic",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
@@ -941,21 +856,15 @@
        "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
-       "entityRanges": Array [
-         Object {
-           "key": 16,
-           "length": 1,
-           "offset": 0,
-         },
-       ],
+       "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "63vue",
-       "text": "ðŸ“·",
-       "type": "unstyled",
+       "text": " ",
+       "type": "atomic",
      },
      Object {
        "data": Object {},
        "depth": 0,
        "entityRanges": Array [],
@@ -1004,17 +913,11 @@
        "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
-       "entityRanges": Array [
-         Object {
-           "key": 17,
-           "length": 6,
-           "offset": 0,
-         },
-       ],
+       "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "f4vnr",
        "text": "[T.C1]Comment",
        "type": "unstyled",
      },
@@ -1028,186 +931,50 @@
        "type": "unstyled",
      },
      Object {
        "data": Object {},
        "depth": 0,
-       "entityRanges": Array [
-         Object {
-           "key": 18,
-           "length": 6,
-           "offset": 0,
-         },
-       ],
+       "entityRanges": Array [],
        "inlineStyleRanges": Array [],
        "key": "6siqr",
        "text": "[T.C2]Comment",
        "type": "unstyled",
      },
    ],
    "entityMap": Object {
      "0": Object {
        "data": Object {
-         "href": "http://www.example.com",
          "url": "http://www.example.com/",
        },
        "mutability": "MUTABLE",
        "type": "LINK",
      },
      "1": Object {
        "data": Object {
-         "href": "mailto:test@example.com",
          "url": "mailto:test@example.com",
        },
        "mutability": "MUTABLE",
        "type": "LINK",
      },
-     "10": Object {
-       "data": Object {
-         "alt": "*",
-         "height": "15",
-         "src": "file:///C:\\\\truncated\\\\msohtmlclip1\\\\01\\\\clip_image001.gif",
-         "width": "15",
-       },
-       "mutability": "MUTABLE",
-       "type": "IMAGE",
-     },
-     "11": Object {
-       "data": Object {
-         "height": "171",
-         "src": "file:///C:\\\\truncated\\\\msohtmlclip1\\\\01\\\\clip_image002.jpg",
-         "width": "258",
-       },
-       "mutability": "MUTABLE",
-       "type": "IMAGE",
-     },
-     "12": Object {
-       "data": Object {
-         "height": "326",
-         "src": "file:///C:\\\\truncated\\\\msohtmlclip1\\\\01\\\\clip_image004.png",
-         "width": "330",
-       },
-       "mutability": "MUTABLE",
-       "type": "IMAGE",
-     },
-     "13": Object {
-       "data": Object {
-         "height": "39",
-         "src": "file:///C:\\\\truncated\\\\msohtmlclip1\\\\01\\\\clip_image006.png",
-         "width": "119",
-       },
-       "mutability": "MUTABLE",
-       "type": "IMAGE",
-     },
-     "14": Object {
-       "data": Object {
-         "height": "97",
-         "src": "file:///C:\\\\truncated\\\\msohtmlclip1\\\\01\\\\clip_image007.png",
-         "width": "195",
-       },
-       "mutability": "MUTABLE",
-       "type": "IMAGE",
-     },
-     "15": Object {
-       "data": Object {
-         "height": "145",
-         "src": "file:///C:\\\\truncated\\\\msohtmlclip1\\\\01\\\\clip_image009.png",
-         "width": "149",
-       },
-       "mutability": "MUTABLE",
-       "type": "IMAGE",
-     },
-     "16": Object {
-       "data": Object {
-         "height": "156",
-         "src": "file:///C:\\\\truncated\\\\msohtmlclip1\\\\01\\\\clip_image011.png",
-         "width": "174",
-       },
-       "mutability": "MUTABLE",
-       "type": "IMAGE",
-     },
-     "17": Object {
-       "data": Object {
-         "href": "#_msoanchor_1",
-         "url": "http://localhost/examples/#_msoanchor_1",
-       },
-       "mutability": "MUTABLE",
-       "type": "LINK",
-     },
-     "18": Object {
-       "data": Object {
-         "href": "#_msoanchor_2",
-         "url": "http://localhost/examples/#_msoanchor_2",
-       },
-       "mutability": "MUTABLE",
-       "type": "LINK",
-     },
      "2": Object {
        "data": Object {
-         "href": "http://www.example.com",
-         "title": "Test tooltip",
          "url": "http://www.example.com/",
        },
        "mutability": "MUTABLE",
        "type": "LINK",
      },
      "3": Object {
        "data": Object {
-         "href": "#_Inline_styles",
-         "url": "http://localhost/examples/#_Inline_styles",
+         "url": "http://www.example.com/",
        },
        "mutability": "MUTABLE",
        "type": "LINK",
      },
      "4": Object {
        "data": Object {
-         "href": "http://www.example.com/",
          "url": "http://www.example.com/",
        },
        "mutability": "MUTABLE",
        "type": "LINK",
      },
-     "5": Object {
-       "data": Object {
-         "href": "#_msocom_1",
-         "url": "http://localhost/examples/#_msocom_1",
    },
-       "mutability": "MUTABLE",
-       "type": "LINK",
-     },
-     "6": Object {
-       "data": Object {
-         "href": "http://www.example.com",
-         "url": "http://www.example.com/",
-       },
-       "mutability": "MUTABLE",
-       "type": "LINK",
-     },
-     "7": Object {
-       "data": Object {
-         "href": "#_msocom_2",
-         "url": "http://localhost/examples/#_msocom_2",
-       },
-       "mutability": "MUTABLE",
-       "type": "LINK",
-     },
-     "8": Object {
-       "data": Object {
-         "alt": "*",
-         "height": "15",
-         "src": "file:///C:\\\\truncated\\\\msohtmlclip1\\\\01\\\\clip_image001.gif",
-         "width": "15",
-       },
-       "mutability": "MUTABLE",
-       "type": "IMAGE",
-     },
-     "9": Object {
-       "data": Object {
-         "alt": "*",
-         "height": "15",
-         "src": "file:///C:\\\\truncated\\\\msohtmlclip1\\\\01\\\\clip_image001.gif",
-         "width": "15",
-       },
-       "mutability": "MUTABLE",
-       "type": "IMAGE",
-     },
-   },
  }
`;
